text/0015-rkv.md

```diff
@@ -0,0 +1,448 @@
+
+### LSM tree LevelDB-alikes derivatives (RocksDB (Facebook), HyperLevelDB, Basho's LevelDB fork etc.)
+
+Dgraph's [summary](https://blog.dgraph.io/post/badger-lmdb-boltdb/):
+
+> LSM tree based stores provide high write performance by sequentially writing key-value pairs in memory in the foreground, and then arranging them in multi-tiered levels on disk in the background. This approach is not without tradeoffs though. Values are written multiple times when arranging them in the tree (known as write amplification) and a single read might need to read multiple levels in the LSM tree before finding a value (known as read amplification).
+
+These various libraries all aim to be faster or more scalable than LevelDB, typically by improving thread parallelism or altering how compaction works (*e.g.*, RocksDB [doesn't use level-based compaction](https://news.ycombinator.com/item?id=6737905)). All are targeted at server workloads.
+
+### Bolt
+
+[A Go implementation of LMDB](https://github.com/boltdb/bolt). B+tree. **Not multiprocess**.
+
+### MDBM
+
+A memory-mapped hash/cache (it would be a stretch to call it persistent) implemented in C, with C++ and Perl bindings. Loses data by default (manual sync), not resistant to power loss/crash, vulnerable to data corruption. No transactions. Keys aren't sorted, so no range queries or cursors. Building your own persistent storage/log is recommended.
+
+[https://github.com/yahoo/mdbm](https://github.com/yahoo/mdbm)
+
+[https://news.ycombinator.com/item?id=8732891](https://news.ycombinator.com/item?id=8732891)
+
+### Kyoto Cabinet
+
+Reportedly very slow. Not multiprocess.
+
+### BerkeleyDB
+
+Very long pedigree. For our purposes, essentially obsoleted in every way by LMDB.
+
+### LMDB
+
+A distillation of experience with BerkeleyDB and use in OpenLDAP. Implemented in C.
+
+Dgraph gives an excellent brief [summary](https://blog.dgraph.io/post/badger-lmdb-boltdb/):
+
+> LMDB provides a key-value stored using[ B+ trees](https://en.wikipedia.org/wiki/B%2B_tree). It has ACID semantics and support for transactions. It does not do any background work and has a crash resilient design that uses[ MVCC](https://en.wikipedia.org/wiki/Multiversion_concurrency_control) instead of locking. This means readers operate on an isolated snapshot of the database and don&rsquo;t block. The codebase is quite small and portable, so LMDB runs across multiple platforms including Android, BSD, and Solaris. This[ talk by Howard Chu](https://www.youtube.com/watch?v=tEa5sAh-kVk&amp;t=10s) at Databaseology 2015 goes into much more details about LMDB&rsquo;s design.
```

Comment:
Nits:
* using[ B+ trees] -> using [B+ trees]
* uses[ MVCC] -> uses [MVCC]
* This[ talk by Howard Chu] -> This [talk by Howard Chu]

Three such typos suggests a problem with the machine translation, although they don't appear elsewhere in the document, only in this paragraph.